### ISSUE TITLE
Type mismatch in function arguments E0631, E0271 are falsely recognized as E0308 mismatched types

### ISSUE THREAD

**Issue by vandenheuvel** [rust-lang/rust#74400]
Regression from `1.45.0-beta.4 (2020-07-13 e99e6422a2f799fc86a3)` to `1.46.0-nightly (2020-07-15 7e11379f3b4c376fbb9a)`.

The following code yields an incorrect error:
```rust
use std::convert::identity;

pub fn is_sorted_and_has_duplicates_by<T, S: PartialEq + PartialOrd>(
    data: &[T],
    key: impl Fn(&T) -> S,
) -> Option<&T> {
    unimplemented!()
}

pub fn is_sorted_and_has_duplicates<T: PartialEq + PartialOrd>(data: &[T]) -> Option<&T> {
    is_sorted_and_has_duplicates_by(data, identity)
}
```
On beta, this gives:
```
error[E0631]: type mismatch in function arguments
  --> src/lib.rs:11:43
   |
3  | pub fn is_sorted_and_has_duplicates_by<T, S: PartialEq + PartialOrd>(
   |        ------------------------------- required by a bound in this
4  |     data: &[T],
5  |     key: impl Fn(&T) -> S,
   |               ----------- required by this bound in `is_sorted_and_has_duplicates_by`
...
11 |     is_sorted_and_has_duplicates_by(data, identity)
   |                                           ^^^^^^^^
   |                                           |
   |                                           expected signature of `for<'r> fn(&'r T) -> _`
   |                                           found signature of `fn(_) -> _`

error[E0271]: type mismatch resolving `for<'r> <fn(_) -> _ {std::convert::identity::<_>} as std::ops::FnOnce<(&'r T,)>>::Output == _`
  --> src/lib.rs:11:5
   |
3  | pub fn is_sorted_and_has_duplicates_by<T, S: PartialEq + PartialOrd>(
   |        ------------------------------- required by a bound in this
4  |     data: &[T],
5  |     key: impl Fn(&T) -> S,
   |                         - required by this bound in `is_sorted_and_has_duplicates_by`
...
11 |     is_sorted_and_has_duplicates_by(data, identity)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected bound lifetime parameter, found concrete lifetime
```

On nightly, this:
```
error[E0308]: mismatched types
  --> src/lib.rs:11:5
   |
11 |     is_sorted_and_has_duplicates_by(data, identity)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected type `std::ops::FnOnce<(&T,)>`
              found type `std::ops::FnOnce<(&T,)>`
```

**Comment by vandenheuvel** [comment]
@jonas-schievink would you be so kind to label this issue? It's sliding down the list while being hard to search.

**Comment by JohnTitor** [comment]
Regression occurred between nightly-2020-06-23 and nightly-2020-06-24, I suspect #73643 is related.

**Comment by LeSeulArtichaut** [comment]
It would be great to find the culprit PR.
@rustbot ping cleanup

**Comment by kanru** [comment]
searched nightlies: from nightly-2020-06-23 to nightly-2020-06-24
regressed nightly: nightly-2020-06-24
searched commits: from https://github.com/rust-lang/rust/commit/6bb3dbfc6c6d8992d08431f320ba296a0c2f7498 to https://github.com/rust-lang/rust/commit/ff5b446d2fdbd898bc97a751f2f72858de185cf1
regressed commit: https://github.com/rust-lang/rust/commit/1557fb031b272b4c5bfcc7de5df7eddc7b36a584

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.5.1</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start=2020-06-23 --end=2020-06-24 --prompt -- check 
```
</details>

If we rule out document changes then we have

#72271
#72493
#73398
#73496
#73515

**Comment by LeSeulArtichaut** [comment]
> Regression occurred between nightly-2020-06-23 and nightly-2020-06-24, I suspect #73643 is related (it's #72493?).
Bisection seems to confirm this, cc @nikomatsakis

**Comment by kanru** [comment]
With a little more git bisect it seems 6873a76f2c34a318a540df65901d07baba1d217b is the first bad commit

**Comment by spastorino** [comment]
Assigning `P-high` as [discussed as part of the Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/I-prioritize.20.2374400.20Type.20mismatch.20in.20function.20arguments.20E063) and removing `I-prioritize`.

**Comment by nikomatsakis** [comment]
Well, I can certainly imagine how this could come about as a result of the leak-check moving, since the same error can now be detected at a different point during type-checking. 

What's happening in that confusing error, I imagine, is that we are now winding up with a "universe error" that comes out from region checking. The identity function was inferred to take an argument `identity::<&'a T>` for some `'a`... 

Well, in short, I imagine that the solution here is to improve the error reporting for the new point where the error occurs, but it'll take a bit more digging to give good mentoring instructions into how to go about doing that.

**Comment by spastorino** [comment]
This is now a stable regression.

**Comment by pnkfelix** [comment]
Discussed in T-compiler meeting. Seems related to (or potential duplicate of) #75791 ?

**Comment by vandenheuvel** [comment]
On today's nightly, the error is now:
```
error: implementation of `FnOnce` is not general enough
  --> src/lib.rs:11:5
   |
11 |     is_sorted_and_has_duplicates_by(data, identity)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
   |
   = note: `fn(&'2 T) -> &'2 T {identity::<&'2 T>}` must implement `FnOnce<(&'1 T,)>`, for any lifetime `'1`...
   = note: ...but it actually implements `FnOnce<(&'2 T,)>`, for some specific lifetime `'2`
```
That seems correct, and with the extra lifetime, the snippet compiles:
```
use std::convert::identity;

pub fn is_sorted_and_has_duplicates_by<'a, T, S: PartialEq + PartialOrd>(
    data: &'a [T],
    key: impl Fn(&'a T) -> S,
) -> Option<&T> {
    unimplemented!()
}

pub fn is_sorted_and_has_duplicates<T: PartialEq + PartialOrd>(data: &[T]) -> Option<&T> {
    is_sorted_and_has_duplicates_by(data, identity)
}
```
